libpod/runtime_ctr.go

```diff
@@ -872,3 +874,89 @@ func (r *Runtime) PruneContainers(filterFuncs []ContainerFilter) (map[string]int
 	}
 	return prunedContainers, pruneErrors, nil
 }
+
+func getAvailableMappings() ([]idtools.IDMap, []idtools.IDMap, error) {
+	if !rootless.IsRootless() {
+		uids, gids, err := rootless.GetConfiguredMappings()
+		if err != nil {
+			return nil, nil, errors.Wrapf(err, "cannot read mappings")
+		}
+		return uids, gids, nil
+	}
+
+	uids, gids, err := rootless.GetConfiguredMappings()
+	if err != nil {
+		return nil, nil, errors.Wrapf(err, "cannot read mappings")
+	}
+
+	// We are already inside on the rootless user namespace.
+	// So we need to remap the configured mappings to what is available inside of the userns.
+	totaluid := 0
+	totalgid := 0
+	for _, u := range uids {
+		totaluid += u.Size
+	}
+	for _, g := range gids {
+		totalgid += g.Size
+	}
+
+	u := []idtools.IDMap{{ContainerID: 0, HostID: 1, Size: totaluid}}
+	g := []idtools.IDMap{{ContainerID: 0, HostID: 1, Size: totalgid}}
+	return u, g, nil
+}
+
+// GetAutoUserNS creates an automatic user namespace
+func GetAutoUserNS(size int, imageName string, rt *Runtime) (*storage.IDMappingOptions, error) {
```

Comment:
Why is size signed? Unsigned seems like it would make more sense.

Reply:
I've picked `int` only because it is the same type used by `IDMap` in containers/storage:

```
type IDMap struct {
	ContainerID int `json:"container_id"`
	HostID      int `json:"host_id"`
	Size        int `json:"size"`
}
```

Reply:
I think this should be a function in container storage.  We might want to deal with this in Buildah and definitely will want to deal with this in CRI-O.